## Question to answer

Identify anomalies between Questions 15 and Question 17.

``` sql
select * 
from 20220813_tw_40.QuestionnaireQuestions qq
;

select sqr.*  /*+ MAX_EXECUTION_TIME(10000000) */
from 20220813_tw_40.SubmittedQuestionResponses sqr 
join 20220813_tw_40.QuestionnaireQuestions qq
	on sqr.question_id  = qq.id 
limit 1000000
; 
```

In [2]:

import pandas as pd


In [4]:
question_df = pd.read_csv('/content/drive/MyDrive/QuestionnaireQuestions-202209061907.csv')
question_df.head()

scope_group_id     id  test_design_version_id       label lang  \
0              37  16961                    1772  Question 1   en   
1              37  16961                    1772  Question 1   en   
2              37  16961                    1772  Question 1   en   
3              37  16961                    1772  Question 1   en   
4              37  16961                    1772  Question 1   fr   

                                       question_text  \
0  **Question 1**\n\nHow often do you have access...   
1  **Question 1**\n\nHow often do you have access...   
2  **Question 1**\n\nHow often do you have access...   
3  **Question 1**\n\nHow often do you have access...   
4  **Question 1**\n\nÀ quelle fréquence as-tu acc...   

                                   sub_question_text    response_id  \
0  \na. I have access to the Internet at home to ...  i16961.e11.r0   
1          \nb. I have access to a computer at home.  i16961.e11.r1   
2  \nc. I have access to a tablet (for example, a...  i16961.e11.r2   
3        \nd. I have access to a smartphone at home.  i16961.e11.r3   
4  \na. J’ai accès à l'Internet à la maison pour ...  i16961.e22.r0   

  response_type  test_window_id  
0  select_table              40  
1  select_table              40  
2  select_table              40  
3  select_table              40  
4  select_table              40

In [5]:
response_df = pd.read_csv('/content/drive/MyDrive/SubmittedQuestionResponses-202209061913.csv')
response_df.head()

response_set_id  scope_group_id    taqr_id  test_attempt_id  \
0  r107690442.i16961.e11.r0              37  107690442          3091346   
1  r107690442.i16961.e11.r1              37  107690442          3091346   
2  r107690442.i16961.e11.r2              37  107690442          3091346   
3  r107690442.i16961.e11.r3              37  107690442          3091346   
4  r107691653.i16961.e11.r0              37  107691653          3120577   

   brd_mident  sch_mident  test_session_id  test_panel_id  stage_id  \
0       66095      422035            71900          10338         1   
1       66095      422035            71900          10338         1   
2       66095      422035            71900          10338         1   
3       66095      422035            71900          10338         1   
4       66095      422035            71902          10338         1   

   stage_form_order  ...  is_nr  lang  question_label  is_invalid  \
0               NaN  ...      0    en      Question 1           0   
1               NaN  ...      0    en      Question 1           0   
2               NaN  ...      0    en      Question 1           0   
3               NaN  ...      0    en      Question 1           0   
4               NaN  ...      0    en      Question 1           0   

  is_operational  is_sample  is_questionnaire sub_session  test_window_id  \
0              1          0                 0         NaN              40   
1              1          0                 0         NaN              40   
2              1          0                 0         NaN              40   
3              1          0                 0         NaN              40   
4              1          0                 0         NaN              40   

   assessment_type  
0  junior_quest_en  
1  junior_quest_en  
2  junior_quest_en  
3  junior_quest_en  
4  junior_quest_en  

[5 rows x 30 columns]

In [6]:
#only one question_id - 16961
response_df.question_id.value_counts()

16961    100000
Name: question_id, dtype: int64

In [7]:
response_df.question_label.value_counts()

Question 1    100000
Name: question_label, dtype: int64

In [8]:
#Number of question label are there in question_df and how many times each question label are repeated
question_df.label.value_counts()

Question 15    40
Question 2     20
Question 5     20
Question 1     19
Question 14    16
Question 13    16
Question 12    16
Question 10    16
Question 9     16
Question 8     16
Question 7     16
Question 6     16
Question 4     16
Question 3     16
Question 11    12
Question 17    12
Question 16     8
Name: label, dtype: int64

###To understand the datasets *better*

In [24]:
#size of the question_df
question_df.shape

(291, 10)

In [10]:
question_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 291 entries, 0 to 290
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   scope_group_id          291 non-null    int64 
 1   id                      291 non-null    int64 
 2   test_design_version_id  291 non-null    int64 
 3   label                   291 non-null    object
 4   lang                    291 non-null    object
 5   question_text           291 non-null    object
 6   sub_question_text       291 non-null    object
 7   response_id             291 non-null    object
 8   response_type           291 non-null    object
 9   test_window_id          291 non-null    int64 
dtypes: int64(4), object(6)
memory usage: 22.9+ KB


In [23]:
response_df.shape

(100000, 30)

In [11]:
response_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 30 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   response_set_id         100000 non-null  object 
 1   scope_group_id          100000 non-null  int64  
 2   taqr_id                 100000 non-null  int64  
 3   test_attempt_id         100000 non-null  int64  
 4   brd_mident              100000 non-null  int64  
 5   sch_mident              100000 non-null  int64  
 6   test_session_id         100000 non-null  int64  
 7   test_panel_id           100000 non-null  int64  
 8   stage_id                100000 non-null  int64  
 9   stage_form_order        0 non-null       float64
 10  module_id               0 non-null       float64
 11  test_design_version_id  100000 non-null  int64  
 12  source_panel_id         0 non-null       float64
 13  question_id             100000 non-null  int64  
 14  response_id          

In [12]:
a= []
for column in question_df.columns:
  if column in response_df.columns:
    a.append(column)
print(a)

['scope_group_id', 'test_design_version_id', 'lang', 'response_id', 'test_window_id']


In [14]:
#compare the question 15 and question 17
df_15_17 = question_df.loc[(question_df['label']=='Question 15') | (question_df['label']=='Question 17')]

In [15]:
df_15_17.head()

scope_group_id     id  test_design_version_id        label lang  \
201              37  18712                    1695  Question 15   en   
202              37  18712                    1695  Question 15   en   
203              37  18712                    1695  Question 15   en   
204              37  18712                    1695  Question 15   en   
205              37  18712                    1695  Question 15   en   

                                         question_text   sub_question_text  \
201  **Question 15**\n\nDid you find the following ...      a. Help menu     
202  **Question 15**\n\nDid you find the following ...          b. Listen    
203  **Question 15**\n\nDid you find the following ...    c. Zoom in/out     
204  **Question 15**\n\nDid you find the following ...    d. Line reader     
205  **Question 15**\n\nDid you find the following ...  e. High contrast     

       response_id response_type  test_window_id  
201  i18712.e11.r0  select_table              40  
202  i18712.e11.r1  select_table              40  
203  i18712.e11.r2  select_table              40  
204  i18712.e11.r3  select_table              40  
205  i18712.e11.r4  select_table              40

In [25]:
df_15_17[['test_window_id','scope_group_id','id','label','question_text','lang','sub_question_text','response_type']].groupby(by=['id','label','question_text','lang','sub_question_text','response_type','test_window_id']).count()

scope_group_id
id    label       question_text                                      lang sub_question_text                                  response_type test_window_id                
18712 Question 15 **Question 15**\n\nAs-tu trouvé utiles les fonc... fr   a. Page d’aide                                     select_table  40                           1
                                                                          b. Écouter                                         select_table  40                           1
                                                                          c. Agrandir/Réduire  \n                            select_table  40                           1
                                                                          d. Lecture ligne par ligne                         select_table  40                           1
                                                                          e. Mode à contraste élevé                          select_table  40                           1
                                                                          f. Outils d’annotation\n(p. ex. surligneur, gom... select_table  40                           1
                                                                          g. Calculatrice                                    select_table  40                           1
                                                                          h. Exercices de respiration                        select_table  40                           1
                                                                          i. Hyperliens (p. ex. un lien cliquable qui te ... select_table  40                           1
                  **Question 15**\n\nDid you find the following f... en   a. Help menu                                       select_table  40                           1
                                                                          b. Listen                                          select_table  40                           1
                                                                          c. Zoom in/out                                     select_table  40                           1
                                                                          d. Line reader                                     select_table  40                           1
                                                                          e. High contrast                                   select_table  40                           1
                                                                          f. Drawing tools (for example, highlighter and ... select_table  40                           1
                                                                          g. Calculator                                      select_table  40                           1
                                                                          h. Calm breathing exercise                         select_table  40                           1
                                                                          i. Hyperlinks (for example, a link you click th... select_table  40                           1
18715 Question 17 **Question 17**\n\nDis-nous dans quelle mesure ... fr   \na. Pour moi, la composante Mathématiques du t... select_table  40                           1
                                                                          \nb. Pour moi, la partie Lecture du test était     select_table  40                           1
                                                                          \nc. Pour moi, la partie Écriture du test était    select_table  40                           1
                  **Question 17**\n\nPlease tell us how easy or d... en   \na. To me, the mathematics part of the assessm... select_table  40                           1
                                                                          \

##Answer 
When compared the question 15 and 17


1.   Both question 15 and 17 had two questions with different `id`
2.   Every question is there in two languages english and french
3.  Question 15 is 'Did you find the following features of the online assessment useful?
4.  Question 17 is 'Please tell us how easy or difficult you found the assessment.'




When I tried to join the question_df and response_df like the SQL query I noticed that response only contain question 
1.  It donot contain question 15 and 17
2.  So I couldnt find anomalous data about question 15 and 17


In [22]:
#Join the data response_df and question_df based on the question_id and id
df_join = response_df.merge(question_df,left_on='question_id',right_on='id',how='left',indicator = True)
df_join

response_set_id  scope_group_id_x    taqr_id  \
0       r107690442.i16961.e11.r0                37  107690442   
1       r107690442.i16961.e11.r0                37  107690442   
2       r107690442.i16961.e11.r0                37  107690442   
3       r107690442.i16961.e11.r0                37  107690442   
4       r107690442.i16961.e11.r0                37  107690442   
...                          ...               ...        ...   
799995  r122427123.i16961.e11.r3                37  122427123   
799996  r122427123.i16961.e11.r3                37  122427123   
799997  r122427123.i16961.e11.r3                37  122427123   
799998  r122427123.i16961.e11.r3                37  122427123   
799999  r122427123.i16961.e11.r3                37  122427123   

        test_attempt_id  brd_mident  sch_mident  test_session_id  \
0               3091346       66095      422035            71900   
1               3091346       66095      422035            71900   
2               3091346       66095      422035            71900   
3               3091346       66095      422035            71900   
4               3091346       66095      422035            71900   
...                 ...         ...         ...              ...   
799995          3152598       67105      871140           108052   
799996          3152598       67105      871140           108052   
799997          3152598       67105      871140           108052   
799998          3152598       67105      871140           108052   
799999          3152598       67105      871140           108052   

        test_panel_id  stage_id  stage_form_order  ...     id  \
0               10338         1               NaN  ...  16961   
1               10338         1               NaN  ...  16961   
2               10338         1               NaN  ...  16961   
3               10338         1               NaN  ...  16961   
4               10338         1               NaN  ...  16961   
...               ...       ...               ...  ...    ...   
799995          10338         1               NaN  ...  16961   
799996          10338         1               NaN  ...  16961   
799997          10338         1               NaN  ...  16961   
799998          10338         1               NaN  ...  16961   
799999          10338         1               NaN  ...  16961   

        test_design_version_id_y       label  lang_y  \
0                           1772  Question 1      en   
1                           1772  Question 1      en   
2                           1772  Question 1      en   
3                           1772  Question 1      en   
4                           1772  Question 1      fr   
...                          ...         ...     ...   
799995                      1772  Question 1      en   
799996                      1772  Question 1      fr   
799997                      1772  Question 1      fr   
799998                      1772  Question 1      fr   
799999                      1772  Question 1      fr   

                                            question_text  \
0       **Question 1**\n\nHow often do you have access...   
1       **Question 1**\n\nHow often do you have access...   
2       **Question 1**\n\nHow often do you have access...   
3       **Question 1**\n\nHow often do you have access...   
4       **Question 1**\n\nÀ quelle fréquence as-tu acc...   
...                                                   ...   
799995  **Question 1**\n\nHow often do you have access...   
799996  **Question 1**\n\nÀ quelle fréquence as-tu acc...   
799997  **Question 1**\n\nÀ quelle fréquence as-tu acc...   
799998  **Question 1**\n\nÀ quelle fréquence as-tu acc...   
799999  **Question 1**\n\nÀ quelle fréquence as-tu acc...   

                                        sub_question_text  response_id_y  \
0       \na. I have access to the Internet at home to ...  i16961.e11.r0   
1               \nb. I have access to a computer at home.  i16961.e11.r1   
2   

## Above dataframe is the joined data by applying the SQL Query 
Two dataframes are joined through left join
on response_df.question_id = question_df.id

1. response_df only had one 100000 rows and question_df had 291 rows
2. Joined dataframe have 800000 rows
